HoRNDIS.cpp

```diff
@@ -968,120 +941,132 @@ UInt32 HoRNDIS::outputPacket(mbuf_t packet, void *param) {
+void HoRNDIS::callbackExit() {
+	fCallbackCount--;
+	// Notify the 'disable' that may be waiting for callback count to reach 0:
+	if (fCallbackCount <= 0) {
+		LOG(V_DEBUG, "Notifying last callback exited");
+		getCommandGate()->commandWakeup(&fCallbackCount);
+	}
+}
+
+void HoRNDIS::dataWriteComplete(void *obj, void *param, IOReturn rc, UInt32 transferred) {
 	HoRNDIS	*me = (HoRNDIS *)obj;
 	unsigned long poolIndx = (unsigned long)param;
 	
 	poolIndx = (unsigned long)param;
-	
-	LOG(V_DEBUG, "(rc %08x, poolIndx %ld)", rc, poolIndx);
-	
-	/* Free the buffer, and hand it off to anyone who might be waiting for one. */
-	me->outbufs[poolIndx].inuse = false;
-	IOLockWakeup(me->outbuf_lock, me->outbufs, true);
-	
-	if (rc == kIOReturnSuccess)
+
+	LOG(V_PACKET, "(rc %08x, poolIndx %ld)", rc, poolIndx);
+	// Callback completed. We don't know when/if we launch another one:
+	me->callbackExit();
+
+	// Note, if 'fReadyToTransfer' is false, we shall not go further:
+	// it's a good idea NOT to touch the 'outbufs'.
+	if (isTransferStopStatus(rc) || !me->fReadyToTransfer) {
+		LOG(V_DEBUG, "Data Write Aborted, or ready-to-transfer is cleared.");
 		return;
-	
-	/* Sigh.  Try to clean up. */
-	LOG(V_ERROR, "I/O error: %08x", rc);
-		
-	if (rc != kIOReturnAborted) {
-		rc = me->clearPipeStall(me->fOutPipe);
-		if (rc != kIOReturnSuccess)
-			LOG(V_ERROR, "clear stall failed (trying to continue)");
 	}
-}
 
-IOReturn HoRNDIS::clearPipeStall(IOUSBPipe *thePipe) {
-	IOReturn rc;
-	
-	if (thePipe->GetPipeStatus() != kIOUSBPipeStalled) {
-		LOG(V_ERROR, "pipe not stalled?");
-		return kIOReturnSuccess;
+	if (rc != kIOReturnSuccess) {
+		// Write error. In case of pipe stall,
+		// we shall clear it on the next transmit.
+		LOG(V_ERROR, "I/O error: %08x", rc);
+	}
+
+	// Free the buffer: put the index back onto the stack:
+	if (me->numFreeOutBufs >= N_OUT_BUFS) {
+		LOG(V_ERROR, "BUG: more free buffers than was allocated");
+		return;
+	}
+
+	me->outbufStack[me->numFreeOutBufs] = poolIndx;
+	me->numFreeOutBufs++;
+	// Unstall the queue whenever the number of free buffers goes 0->1.
+	// I.e. we unstall it the moment we're able to write something into it:
+	if (me->numFreeOutBufs == 1) {
+		me->getOutputQueue()->service();
 	}
-	
-	rc = thePipe->ClearPipeStall(true);
-	LOG(V_ERROR, "pipe stall clear: rv %08x", rc);
-	
-	return rc;
 }
 
 
 /***** Packet receive logic *****/
-
-void HoRNDIS::dataReadComplete(void *obj, void *param, IOReturn rc, UInt32 remaining) {
+void HoRNDIS::dataReadComplete(void *obj, void *param, IOReturn rc, UInt32 transferred) {
 	HoRNDIS	*me = (HoRNDIS *)obj;
+	pipebuf_t *inbuf = (pipebuf_t *)param;
 	IOReturn ior;
-	
-	if (rc == kIOReturnAborted || rc == kIOReturnNotResponding) {
-		LOG(V_ERROR, "I/O aborted: device unplugged?");
+
+	// Stop conditions. Not separating them out, since reacting to individual
+	// ones would be very timing-sansitive.
+	if (isTransferStopStatus(rc) || !me->fReadyToTransfer) {
```

Comment:
Unusual cases probably deserve more than V_DEBUG (unless this happens on every unplug)?

Reply:
Yep, that happens every time you unplug the device or disable the tethering in the Android settings.